### PR TITLE
IGNITE-17221 HLC propagating with RAFT events.

### PR DESCRIPTION
**Pull request by sergeyuttsel**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java [hunk 1]
```diff
@@ -17,13 +17,14 @@
...
 
+import java.io.Serializable;
 import org.apache.ignite.internal.tostring.S;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * A hybrid timestamp that combines physical clock and logical clock.
  */
-public class HybridTimestamp implements Comparable<HybridTimestamp> {
+public class HybridTimestamp implements Comparable<HybridTimestamp>, Serializable {
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java:27]: This is inefficient and error prone due to missing serialVersionUID. Can we made it a message or use manual serialization ?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 2]
```diff
@@ -541,6 +547,15 @@ public NodeImpl(final String groupId, final PeerId serverId) {
...
         this.confCtx = new ConfigurationCtx(this);
         this.wakingCandidate = null;
+        this.clock = clock;
+    }
+
+    public HybridTimestamp clockNow() {
+        return clock.now();
+    }
+
+    public HybridTimestamp clockTick(HybridTimestamp timestamp) {
```
Comment (ascherbakoff) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:557]: Redundant methods

File: modules/core/src/main/java/org/apache/ignite/hlc/SystemTimeProvider.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.apache.ignite.hlc;
+
+import java.time.Clock;
+
+/**
+ * Time provider which use system time.
+ */
+public class SystemTimeProvider implements PhysicalTimeProvider {
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/hlc/SystemTimeProvider.java:25]: Any reason to have more than one instance of SystemTimeProvider ?

File: modules/core/src/main/java/org/apache/ignite/hlc/HybridClock.java [hunk 4]
```diff
@@ -17,59 +17,22 @@
...
-    private HybridTimestamp latestTime;
-
-    /**
-     * The constructor which initializes the latest time to current time by system clock.
-     */
-    public HybridClock() {
-        this.latestTime = new HybridTimestamp(Clock.systemUTC().instant().toEpochMilli(), 0);
-    }
-
+public interface HybridClock {
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/hlc/HybridClock.java:23]: Any reason to have this interface ? I can't imaging more than one implementation.

File: modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java]: Fixed. I removed Serializable.

File: modules/core/src/main/java/org/apache/ignite/hlc/HybridClock.java [hunk 5]
```diff
@@ -17,59 +17,22 @@
...
-
-        return latestTime;
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public String toString() {
-        return S.toString(HybridClock.class, this);
-    }
+    HybridTimestamp tick(HybridTimestamp requestTime);
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/hlc/HybridClock.java]: Ok. Fixed.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 6]
```diff
@@ -529,7 +535,7 @@ boolean isBusy() {
         }
     }
 
-    public NodeImpl(final String groupId, final PeerId serverId) {
+    public NodeImpl(final String groupId, final PeerId serverId, HybridClock clock) {
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java]: Fixed. I used DefaultJRaftServiceFactory and NodeOptions.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java (hunk 2 above)
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java]: Why? It is used.

File: modules/core/src/main/java/org/apache/ignite/hlc/SystemTimeProvider.java (hunk 3 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/hlc/SystemTimeProvider.java]: Fixed. Seems that it is not need now.

File: modules/core/src/main/java/org/apache/ignite/hlc/HybridClock.java (hunk 4 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/hlc/HybridClock.java]: Fixed. Seems that it doesn't need now.

File: modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java (hunk 1 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/hlc/HybridClockTest.java [hunk 7]
```diff
@@ -48,7 +48,7 @@ public void afterEach() {
     public void testNow() {
         clockMock = mockToEpochMilli(100);
 
-        HybridClock clock = new HybridClock();
+        HybridClock clock = new HybridClock(new SystemTimeProvider());
```
Comment (ascherbakoff) [inline modules/core/src/test/java/org/apache/ignite/hlc/HybridClockTest.java:51]: Passing a specific implementation looks really weird to me. Do you expect other implementations to extend SystemTimeProvider? This doesn't look correct. Why not to implement getter as a closure and pass it like new HybridClock(SystemTimeProvider::getPhysicalTime)

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/JRaftServiceFactory.java [hunk 8]
```diff
@@ -57,6 +58,13 @@ public interface JRaftServiceFactory {
      */
     RaftMetaStorage createRaftMetaStorage(final String uri, final RaftOptions raftOptions);
 
+    /**
+     * A hybrid logical clock.
+     *
+     * @return A hybrid logical clock.
+     */
+    HybridClock getHybridClock();
```
Comment (ascherbakoff) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/JRaftServiceFactory.java:66]:
```suggestion
    HybridClock getClock();
```

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 9]
```diff
@@ -96,11 +98,16 @@ public class Loza implements IgniteComponent {
...
+     * @param clock         A hybrid logical clock.
      */
-    public Loza(ClusterService clusterNetSvc, Path dataPath) {
+    public Loza(ClusterService clusterNetSvc, Path dataPath, HybridClock clock) {
         this.clusterNetSvc = clusterNetSvc;
 
-        this.raftServer = new JraftServerImpl(clusterNetSvc, dataPath);
+        NodeOptions options = new NodeOptions();
+
+        options.setClock(clock);
```
Comment (ascherbakoff) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:108]: We already have this in service factory

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 10]
```diff
@@ -1951,11 +1971,12 @@ else if (candidateId.equals(leaderId)) { // Already follows a leader in this ter
             }
             while (false);
 
-            return raftOptions.getRaftMessagesFactory()
-                .requestVoteResponse()
-                .term(this.currTerm)
-                .granted(request.term() == this.currTerm && candidateId.equals(this.votedId))
-                .build();
+            RequestVoteResponseBuilder rb = raftOptions.getRaftMessagesFactory()
```
Comment (ascherbakoff) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:1974]: Unncecessary change

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/Replicator.java [hunk 11]
```diff
@@ -1535,6 +1541,7 @@ void sendEntries() {
      */
     private boolean sendEntries(final long nextSendingIndex) {
         final AppendEntriesRequestBuilder rb = raftOptions.getRaftMessagesFactory().appendEntriesRequest();
+        rb.timestamp(options.getNode().clockNow());
```
Comment (ascherbakoff) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/Replicator.java:1544]: We have agreed to send timestamps only in heartbeats

File: modules/core/src/test/java/org/apache/ignite/hlc/HybridClockTest.java (hunk 7 above)
Comment (sergeyuttsel) [reply modules/core/src/test/java/org/apache/ignite/hlc/HybridClockTest.java]: I removed a time provider. I created it for tests but it is not needed now.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/JRaftServiceFactory.java (hunk 8 above)
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/JRaftServiceFactory.java]: Fixed.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java (hunk 9 above)
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:108]: It is needed to pass clock to JraftServerImpl.
JraftServerImpl creates DefaultJRaftServiceFactory with the clock instance.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java (hunk 10 above)
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java]: Fixed here and other similar places.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/Replicator.java (hunk 11 above)
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/Replicator.java]: Yes, there is no need to use a clock here. Fixed.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java (hunk 9 above)
Comment (sergeyuttsel) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:108]: I merged main into ignite3_tx and ignite-17221.
Now the clock is passed by NodeOptions.

**Pull request merged**